src/main/kotlin/com/openlattice/organizations/HazelcastOrganizationService.kt

```diff
@@ -285,44 +290,47 @@ class HazelcastOrganizationService(
         }
     }
 
+    /**
+     * Do not invoke this directly as it bypasses checks in [HazelcastOrganizationService.addMembers]
+     *
+     */
     private fun addMembers(
             orgAclKey: AclKey,
-            membersToAdd: Set<Principal>,
+            membersToAdd: Map<Principal, AclKey>,
             profiles: Map<Principal, Map<String, Set<String>>>
     ): Set<Principal> {
         require(orgAclKey.size == 1) { "Organization acl key should only be of length 1" }
-
-        val members = if (membersToAdd is Serializable) membersToAdd else membersToAdd.toSet()
-
-        val nonUserPrincipals = members.filter { it.type != PrincipalType.USER }
-        require(nonUserPrincipals.isEmpty()) { "Cannot add non-users principals $nonUserPrincipals to organization $orgAclKey" }
+        val members = membersToAdd.keys.toSet()
         val organizationId = orgAclKey[0]
-
-
         val newMembers: Set<Principal> = organizations.executeOnKey(organizationId, OrganizationEntryProcessor {
-            val newMembers = members.filter { member -> !it.members.contains(member) }.toSet()
+            val newMembers = members - it.members
             it.members.addAll(newMembers)
-            return@OrganizationEntryProcessor Result(newMembers, newMembers.isNotEmpty())
+            //We're always persisting here.
+            return@OrganizationEntryProcessor Result(newMembers)
         }) as Set<Principal>
 
-        if (newMembers.isNotEmpty()) {
-            grantOrganizationPrincipals(organizationId, newMembers, orgAclKey, profiles)
-        }
+        //Always trigger as this won't cause a write to organizations table.
+        grantOrganizationPrincipals(
+                organizationId,
+                membersToAdd,
+                orgAclKey,
```

Comment:
Silly question... is the `orgAclKey` param really necessary? Isn't the `organizationId` param enough info?

Reply:
It's a compatibility thing for callers. You don't have to provide the parameter as it will be automatically instantiated from the provided organizationId. It can be useful to avoid making an unnecessary copy if the caller has the organization acl key around for some reason.